tcpdirect.cu

```diff
@@ -0,0 +1,785 @@
+#ifdef WITH_TCPDIRECT
+#include <cuda.h>
+#include <cuda_runtime.h>
+
+#include <asm-generic/errno-base.h>
+#include <asm-generic/socket.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <linux/if.h>
+#include <stdlib.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <sys/socket.h>
+#include <unistd.h>
+#include <linux/dma-buf.h>
+
+#include <memory>
+#include <cstdint>
+#include <cstdio>
+#include <string>
+#include <vector>
+
+#include "tcpdirect.h"
+#include "logging.h"
+#include "flow.h"
+#include "thread.h"
+
+#ifndef MSG_ZEROCOPY
+#define MSG_ZEROCOPY	0x4000000
+#endif
+
+#define LAST_PRIME 111
+
+#define MIN_RX_BUFFER_TOTAL_SIZE (1 << 28)
+#define GPUMEM_ALIGNMENT (1UL << 21)
+#define GPUMEM_MINSZ 0x400000
+#define PAGE_SHIFT (12)
+#define PAGE_SIZE (1 << PAGE_SHIFT)
+
+#define multiplier (1 << 16)
+
+#define TEST_PREFIX "ncdevmem"
+#define NUM_PAGES 16000
+
+/* missing definitions in mman-linux.h */
+#ifndef MFD_ALLOW_SEALING
+#define MFD_ALLOW_SEALING 2U
+#endif
+
+/* GRTE libraries from google3 already define the following */
+#ifndef F_SEAL_SHRINK
+#define F_SEAL_SHRINK 2U
+#endif
+#ifndef F_ADD_SEALS
+#define F_ADD_SEALS 1033U
+#endif
+#ifndef F_GET_SEALS
+#define F_GET_SEALS 1034U
+#endif
```

Comment:
MFD_ALLOW or F_SEAL defines. Are you sure you need them to compile?

Adding explicit here is actually a bit ugly. Technically they should come from the kernel headers. Try removing them, and see what the compile error is.